```
Error running mint automation
+ mkdir -p 14094-4e35a70/gopath/src/github.com/minio
+ git -C 14094-4e35a70/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 14094-4e35a70/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 14094-4e35a70/gopath/src/github.com/minio/minio fetch --quiet minio pull/14094/head:pr14094
+ git -C 14094-4e35a70/gopath/src/github.com/minio/minio checkout --quiet pr14094
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/14094-4e35a70/gopath
+ make -C 14094-4e35a70/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/colinmarc/hdfs/v2 v2.2.0
go: downloading github.com/minio/madmin-go v1.2.4
go: downloading github.com/minio/cli v1.22.0
go: downloading github.com/minio/minio-go/v7 v7.0.20
go: downloading github.com/minio/pkg v1.1.14
go: downloading google.golang.org/api v0.58.0
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/minio/kes v0.14.0
go: downloading github.com/minio/sio v0.3.0
go: downloading golang.org/x/sys v0.0.0-20211020174200-9d6173849985
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/Shopify/sarama v1.27.2
go: downloading github.com/rs/dnscache v0.0.0-20210201191234-295bba877686
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading golang.org/x/crypto v0.0.0-20211209193657-4570a0811e8b
go: downloading github.com/klauspost/compress v1.13.6
go: downloading github.com/golang-jwt/jwt/v4 v4.1.0
go: downloading github.com/VividCortex/ewma v1.1.1
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/coreos/go-oidc v2.1.0+incompatible
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/miekg/dns v1.1.43
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/klauspost/reedsolomon v1.9.15
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/go-openapi/loads v0.20.2
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/klauspost/readahead v1.3.1
go: downloading github.com/minio/console v0.13.2
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/minio/selfupdate v0.4.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.0+incompatible
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/prometheus/client_golang v1.11.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/shirou/gopsutil/v3 v3.21.9
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.0
go: downloading go.etcd.io/etcd/client/v3 v3.5.0
go: downloading go.uber.org/atomic v1.9.0
go: downloading golang.org/x/oauth2 v0.0.0-20210819190943-2bc19b11175f
go: downloading github.com/coredns/coredns v1.4.0
go: downloading github.com/go-ldap/ldap/v3 v3.2.4
go: downloading golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac
go: downloading go.uber.org/zap v1.19.1
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/minio/simdjson-go v0.2.1
go: downloading github.com/eclipse/paho.mqtt.golang v1.3.5
go: downloading github.com/golang-jwt/jwt v3.2.2+incompatible
go: downloading github.com/elastic/go-elasticsearch/v7 v7.12.0
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/gomodule/redigo v1.8.8
go: downloading github.com/lib/pq v1.9.0
go: downloading github.com/nats-io/nats.go v1.11.1-0.20210623165838-4b75fc59ae30
go: downloading github.com/nats-io/stan.go v0.8.3
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/pquerna/cachecontrol v0.0.0-20171018203845-0dec1b30a021
go: downloading gopkg.in/square/go-jose.v2 v2.3.1
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/mattn/go-colorable v0.1.10
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/minio/parquet-go v1.1.0
go: downloading github.com/klauspost/cpuid/v2 v2.0.9
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading golang.org/x/net v0.0.0-20211112202133-69e39bad7dc2
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20200615164410-66371956d46c
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.1
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.0
go: downloading google.golang.org/grpc v1.41.0
go: downloading github.com/golang/snappy v0.0.3
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/prometheus/common v0.31.1
go: downloading github.com/beorn7/perks v1.0.1
go: downloading cloud.google.com/go v0.94.1
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/gorilla/websocket v1.4.2
go: downloading google.golang.org/genproto v0.0.0-20210928142010-c7af6a1a74c9
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/go-openapi/analysis v0.20.0
go: downloading github.com/go-openapi/swag v0.19.14
go: downloading github.com/go-openapi/spec v0.20.3
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/minio/argon2 v1.0.0
go: downloading go.uber.org/multierr v1.7.0
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading google.golang.org/protobuf v1.27.1
go: downloading github.com/rs/xid v1.3.0
go: downloading gopkg.in/ini.v1 v1.63.2
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/tklauser/go-sysconf v0.3.9
go: downloading github.com/lestrrat-go/jwx v1.2.7
go: downloading github.com/fatih/structs v1.1.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20200313005456-10cdbea86bc0
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.5.0
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/mailru/easyjson v0.7.6
go: downloading github.com/jcmturner/goidentity/v6 v6.0.1
go: downloading github.com/tklauser/numcpus v0.3.0
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/tidwall/sjson v1.2.3
go: downloading github.com/tidwall/gjson v1.11.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/go-openapi/strfmt v0.20.0
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/mitchellh/mapstructure v1.4.1
go: downloading github.com/asaskevich/govalidator v0.0.0-20200907205600-7a23bdc65eef
go: downloading github.com/go-openapi/errors v0.19.9
go: downloading go.mongodb.org/mongo-driver v1.4.6
go: downloading github.com/go-openapi/jsonreference v0.19.5
go: downloading github.com/PuerkitoBio/purell v1.1.1
go: downloading github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
go: downloading github.com/googleapis/gax-go/v2 v2.1.1
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
go: downloading github.com/go-stack/stack v1.8.0
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
go: downloading github.com/go-openapi/runtime v0.19.24
go: downloading github.com/jessevdk/go-flags v1.4.0
go: downloading github.com/minio/operator/logsearchapi v0.0.0-20211011212245-31460bbbc4b7
go: downloading github.com/minio/operator v0.0.0-20211011212245-31460bbbc4b7
go: downloading github.com/minio/mc v0.0.0-20211207230606-23a05f5a17f2
go: downloading github.com/unrolled/secure v1.0.9
go: downloading github.com/minio/direct-csi v1.3.5-0.20210601185811-f7776f7961bf
go: downloading k8s.io/client-go v0.21.1
go: downloading github.com/go-openapi/validate v0.20.2
go: downloading github.com/docker/go-units v0.4.0
go: downloading k8s.io/apimachinery v0.21.1
go: downloading github.com/mb0/glob v0.0.0-20160210091149-1eb79d2de6c4
go: downloading k8s.io/kube-openapi v0.0.0-20210305001622-591a79e4bda7
go: downloading github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
go: downloading maze.io/x/duration v0.0.0-20160924141736-faac084b6075
go: downloading github.com/minio/colorjson v1.0.1
go: downloading github.com/pkg/xattr v0.4.3
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/briandowns/spinner v1.16.0
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/posener/complete v1.2.3
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/pkg/profile v1.6.0
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading k8s.io/klog/v2 v2.8.0
go: downloading k8s.io/api v0.21.1
go: downloading github.com/georgysavva/scany v0.2.7
go: downloading sigs.k8s.io/controller-runtime v0.8.0
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/go-logr/logr v0.4.0
go: downloading github.com/stretchr/testify v1.7.0
go: downloading github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2
go: downloading github.com/stretchr/objx v0.2.0
go: downloading gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
go: downloading sigs.k8s.io/yaml v1.2.0
go: downloading github.com/google/gofuzz v1.1.0
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.1.0
go: downloading github.com/google/go-cmp v0.5.6
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading github.com/emicklei/go-restful v2.9.5+incompatible
go: downloading k8s.io/utils v0.0.0-20201110183641-67b214c5f920
go: downloading github.com/googleapis/gnostic v0.5.1
# github.com/minio/minio/cmd
cmd/erasure-decode.go:325:7: assignment mismatch: 1 variable but w.Write returns 2 values
make: *** [build] Error 2

```

--- a/cmd/bitrot.go
+++ b/cmd/bitrot.go
@@ -119,8 +119,10 @@ func newBitrotReader(disk StorageAPI, data []byte, bucket string, filePath strin
 // Close all the readers.
 func closeBitrotReaders(rs []io.ReaderAt) {
 	for _, r := range rs {
-		if br, ok := r.(io.Closer); ok {
-			br.Close()
+		if r != nil {
+			if br, ok := r.(io.Closer); ok {
+				br.Close()
+			}
 		}
 	}
 }
@@ -128,8 +130,10 @@ func closeBitrotReaders(rs []io.ReaderAt) {
 // Close all the writers.
 func closeBitrotWriters(ws []io.Writer) {
 	for _, w := range ws {
-		if bw, ok := w.(io.Closer); ok {
-			bw.Close()
+		if w != nil {
+			if bw, ok := w.(io.Closer); ok {
+				bw.Close()
+			}
 		}
 	}
 }

--- a/cmd/erasure-decode.go
+++ b/cmd/erasure-decode.go
@@ -322,8 +322,8 @@ func (e Erasure) Heal(ctx context.Context, writers []io.Writer, readers []io.Rea
 			errs:        make([]error, len(writers)),
 		}
 
-		err = w.Write(ctx, bufs)
-		if err != nil {
+		if err = w.Write(ctx, bufs); err != nil {
+			logger.LogIf(ctx, err)
 			return err
 		}
 	}

--- a/cmd/erasure-encode.go
+++ b/cmd/erasure-encode.go
@@ -52,7 +52,10 @@ func (p *parallelWriter) Write(ctx context.Context, blocks [][]byte) error {
 			if p.errs[i] == nil {
 				if n != len(blocks[i]) {
 					p.errs[i] = io.ErrShortWrite
+					p.writers[i] = nil
 				}
+			} else {
+				p.writers[i] = nil
 			}
 		}(i)
 	}
@@ -93,6 +96,7 @@ func (e *Erasure) Encode(ctx context.Context, src io.Reader, writers []io.Writer
 		blocks, err = e.EncodeData(ctx, buf[:n])
 		if err != nil {
 			logger.LogIf(ctx, err)
+
 			return 0, err
 		}
 

--- a/cmd/erasure-healing.go
+++ b/cmd/erasure-healing.go
@@ -540,7 +540,7 @@ func (er erasureObjects) healObject(ctx context.Context, bucket string, object s
 
 			// If all disks are having errors, we give up.
 			if disksToHealCount == 0 {
-				return result, fmt.Errorf("all disks had write errors, unable to heal")
+				return result, fmt.Errorf("all disks had write errors, unable to heal %s/%s", bucket, object)
 			}
 
 		}

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -701,9 +701,10 @@ func (er erasureObjects) putMetacacheObject(ctx context.Context, key string, r *
 		if disk == nil {
 			continue
 		}
-
-		inlineBuffers[i] = bytes.NewBuffer(make([]byte, 0, shardFileSize))
-		writers[i] = newStreamingBitrotWriterBuffer(inlineBuffers[i], DefaultBitrotAlgorithm, erasure.ShardSize())
+		if disk.IsOnline() {
+			inlineBuffers[i] = bytes.NewBuffer(make([]byte, 0, shardFileSize))
+			writers[i] = newStreamingBitrotWriterBuffer(inlineBuffers[i], DefaultBitrotAlgorithm, erasure.ShardSize())
+		}
 	}
 
 	n, erasureErr := erasure.Encode(ctx, data, writers, buffer, writeQuorum)
@@ -720,6 +721,7 @@ func (er erasureObjects) putMetacacheObject(ctx context.Context, key string, r *
 
 	for i, w := range writers {
 		if w == nil {
+			// Make sure to avoid writing to disks which we couldn't complete in erasure.Encode()
 			onlineDisks[i] = nil
 			continue
 		}
@@ -930,6 +932,10 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 			continue
 		}
 
+		if !disk.IsOnline() {
+			continue
+		}
+
 		if len(inlineBuffers) > 0 {
 			sz := shardFileSize
 			if sz < 0 {
@@ -939,6 +945,7 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 			writers[i] = newStreamingBitrotWriterBuffer(inlineBuffers[i], DefaultBitrotAlgorithm, erasure.ShardSize())
 			continue
 		}
+
 		writers[i] = newBitrotWriter(disk, minioMetaTmpBucket, tempErasureObj, shardFileSize, DefaultBitrotAlgorithm, erasure.ShardSize())
 	}
 
@@ -1012,6 +1019,9 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 
 	// Rename the successfully written temporary object to final location.
 	if onlineDisks, err = renameData(ctx, onlineDisks, minioMetaTmpBucket, tempObj, partsMetadata, bucket, object, writeQuorum); err != nil {
+		if errors.Is(err, errFileNotFound) {
+			return ObjectInfo{}, toObjectErr(errErasureWriteQuorum, bucket, object)
+		}
 		logger.LogIf(ctx, err)
 		return ObjectInfo{}, toObjectErr(err, bucket, object)
 	}

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -291,6 +291,7 @@ func configRetriableErrors(err error) bool {
 		errors.Is(err, io.ErrUnexpectedEOF) ||
 		errors.As(err, &rquorum) ||
 		errors.As(err, &wquorum) ||
+		isErrObjectNotFound(err) ||
 		isErrBucketNotFound(err) ||
 		errors.Is(err, os.ErrDeadlineExceeded)
 }

--- a/cmd/xl-storage.go
+++ b/cmd/xl-storage.go
@@ -1984,6 +1984,14 @@ func (s *xlStorage) Delete(ctx context.Context, volume string, path string, recu
 	return s.deleteFile(volumeDir, filePath, recursive)
 }
 
+func skipAccessChecks(volume string) bool {
+	switch volume {
+	case minioMetaTmpBucket, minioMetaBucket, minioMetaMultipartBucket:
+		return true
+	}
+	return false
+}
+
 // RenameData - rename source path to destination path atomically, metadata and data directory.
 func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, fi FileInfo, dstVolume, dstPath string) (err error) {
 	defer func() {
@@ -2010,23 +2018,27 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
 		return err
 	}
 
-	// Stat a volume entry.
-	if err = Access(srcVolumeDir); err != nil {
-		if osIsNotExist(err) {
-			return errVolumeNotFound
-		} else if isSysErrIO(err) {
-			return errFaultyDisk
+	if !skipAccessChecks(srcVolume) {
+		// Stat a volume entry.
+		if err = Access(srcVolumeDir); err != nil {
+			if osIsNotExist(err) {
+				return errVolumeNotFound
+			} else if isSysErrIO(err) {
+				return errFaultyDisk
+			}
+			return err
 		}
-		return err
 	}
 
-	if err = Access(dstVolumeDir); err != nil {
-		if osIsNotExist(err) {
-			return errVolumeNotFound
-		} else if isSysErrIO(err) {
-			return errFaultyDisk
+	if !skipAccessChecks(dstVolume) {
+		if err = Access(dstVolumeDir); err != nil {
+			if osIsNotExist(err) {
+				return errVolumeNotFound
+			} else if isSysErrIO(err) {
+				return errFaultyDisk
+			}
+			return err
 		}
-		return err
 	}
 
 	srcFilePath := pathutil.Join(srcVolumeDir, pathJoin(srcPath, xlStorageFormatFile))
@@ -2237,7 +2249,8 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
 					s.deleteFile(dstVolumeDir, legacyDataPath, true)
 				}
 				s.deleteFile(dstVolumeDir, dstDataPath, false)
-				if err != errFileNotFound {
+				// Looks like srcFilePath is missing usually at .minio.sys/ ignore it.
+				if !errors.Is(err, errFileNotFound) {
 					logger.LogIf(ctx, err)
 				}
 				return osErrToFileErr(err)
@@ -2252,9 +2265,11 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
 			}
 			s.deleteFile(dstVolumeDir, dstFilePath, false)
 
-			if err != errFileNotFound {
+			// Looks like srcFilePath is missing usually at .minio.sys/ ignore it.
+			if !errors.Is(err, errFileNotFound) {
 				logger.LogIf(ctx, err)
 			}
+
 			return osErrToFileErr(err)
 		}
 

